src/main/java/uk/gov/ida/verifyserviceprovider/services/UnifyingKeySignatureTrustEngine.java

```diff
@@ -0,0 +1,105 @@
+import net.shibboleth.utilities.java.support.resolver.CriteriaSet;
+import net.shibboleth.utilities.java.support.resolver.ResolverException;
+import org.opensaml.security.credential.Credential;
+import org.opensaml.security.credential.CredentialResolver;
+import org.opensaml.xmlsec.keyinfo.KeyInfoCredentialResolver;
+import org.opensaml.xmlsec.signature.support.impl.ExplicitKeySignatureTrustEngine;
+import uk.gov.ida.saml.metadata.EidasMetadataResolverRepository;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+
+public class UnifyingKeySignatureTrustEngine extends ExplicitKeySignatureTrustEngine {
+
+    public UnifyingKeySignatureTrustEngine(ExplicitKeySignatureTrustEngine hubTrustEngine, EidasMetadataResolverRepository eidasMetadataResolverRepository) {
+        super(
+            new UnifyingCredentialResolver(hubTrustEngine, eidasMetadataResolverRepository),
+            new UnifyingKeyInfoCredentialResolver(hubTrustEngine, eidasMetadataResolverRepository));
+    }
+
+    private static class UnifyingKeyInfoCredentialResolver implements KeyInfoCredentialResolver {
+
+        private final ExplicitKeySignatureTrustEngine hubTrustEngine;
+        private final EidasMetadataResolverRepository eidasMetadataResolverRepository;
+
+        private UnifyingKeyInfoCredentialResolver(ExplicitKeySignatureTrustEngine hubTrustEngine, EidasMetadataResolverRepository eidasMetadataResolverRepository) {
+            this.hubTrustEngine = hubTrustEngine;
+            this.eidasMetadataResolverRepository = eidasMetadataResolverRepository;
+        }
+
+        @Nonnull
+        @Override
+        public Iterable<Credential> resolve(@Nullable CriteriaSet criteria) throws ResolverException {
+            List<Credential> credentialsFound = new LinkedList<>();
+
+            for (String eidasEntityId : eidasMetadataResolverRepository.getResolverEntityIds()) {
+                KeyInfoCredentialResolver eidasCredentialResolver = eidasMetadataResolverRepository.getSignatureTrustEngine(eidasEntityId).get().getKeyInfoResolver();
+                credentialsFound.addAll(Lists.newLinkedList(eidasCredentialResolver.resolve(criteria)));
+            }
+
+            KeyInfoCredentialResolver hubCredentialResolver = hubTrustEngine.getKeyInfoResolver();
+            credentialsFound.addAll(Lists.newLinkedList(hubCredentialResolver.resolve(criteria)));
+
+            return credentialsFound;
+        }
+
+        @Nullable
+        @Override
+        public Credential resolveSingle(@Nullable CriteriaSet criteria) throws ResolverException {
+            Iterable<Credential> credentials = resolve(criteria);
+            if (credentials.iterator().hasNext()) {
+                Credential credential = credentials.iterator().next();
+                if (credentials.iterator().hasNext()) { throw new ResolverException("More than 1 credential found for criteria provided."); }
```

Comment:
Style: We shouldn't be using one line `if` statements even for simple checks. I know you can find it in the code base but the official style guide we use is not to do this.